types/src/identifier.rs

```diff
@@ -3,18 +3,54 @@
 
 //! An identifier is the name of an entity (module, resource, function, etc) in Move.
 //!
+//! A valid identifier consists of an ASCII string which satisfies any of the conditions:
+//!
+//! * The first character is a letter and the remaining characters are letters, digits or
+//!   underscores.
+//! * The first character is an underscore, and there is at least one further letter, digit or
+//!   underscore.
+//!
+//! The spec for allowed identifiers is similar to Rust's spec
+//! ([as of version 1.38](https://doc.rust-lang.org/1.38.0/reference/identifiers.html)).
+//!
+//! Allowed identifiers are currently restricted to ASCII due to unresolved issues with Unicode
+//! normalization. See [Rust issue #55467](https://github.com/rust-lang/rust/issues/55467) and the
+//! associated RFC for some discussion. Unicode identifiers may eventually be supported once these
+//! issues are worked out.
+//!
+//! This module only determines allowed identifiers at the bytecode level. Move source code will
+//! likely be more restrictive than even this, with a "raw identifier" escape hatch similar to
+//! Rust's `r#` identifiers.
+//!
 //! Among other things, identifiers are used to:
 //! * specify keys for lookups in storage
 //! * do cross-module lookups while executing transactions
 
-// TODO: restrict identifiers to a subset of ASCII
-
 use failure::prelude::*;
+use lazy_static::lazy_static;
 #[cfg(any(test, feature = "testing"))]
 use proptest::prelude::*;
+use regex::Regex;
 use serde::{Deserialize, Serialize};
 use std::{borrow::Borrow, fmt, ops::Deref};
 
+/// Describes what identifiers are allowed.
+///
+/// For now this is deliberately restrictive -- we would like to evolve this in the future.
+// TODO: "<SELF>" is coded as an exception. It should be removed once CompiledScript goes away.
+static ALLOWED_IDENTIFIERS: &str = r"(?:[a-zA-Z][a-zA-Z0-9_]*)|(?:_[a-zA-Z0-9_]+)|(?:<SELF>)";
```

Comment:
So here's how my thinking about this went when I was writing it:

* ugh, regexes feel icky
* ok, how would I go about writing a parser?
* I'd probably handwrite a finite state machine...
* (recalling CS theory: the set of grammars described by regexes (without backcaptures) is exactly the set of grammars described by finite state machines)
* ...but the regex crate already builds finite state machines out of regular expressions
* and the regex crate is already in the TCB
* so why not just use it? it's simpler, easier to review and will lead to the same result.

Reply:
I don't really like the regex either, but I think the regex is probably fine. 

I wrote a non regex one; its  easier to reason about (for me) but maybe a little verbose 
```
fn is_underscore_alpha_or_digit(c: &char) -> bool {
  match c { '_' | 'a'..='z' | 'A'..='Z' => true | _ => false }
}

fn is_valid(s: &str) -> bool {
  if s == "<SELF>" { return true;  }
  let len = s.len();
  let chars = s.chars();
  match chars.next() {
     Some('a'..='z') |
     Some('A'..='Z') => chars.all(is_underscore_alpha_or_digit), 
     Some('_') if len > 1 => chars.all(is_underscore_alpha_or_digit), 
     _ => false
  }
```

Reply:
I agree with @tnowacki about this being very easy and the code he has is more or less what I would have had.
I don't know that the argument of "building a finite state machine and regex doing it already" makes a lot of sense to me (and it's me of course, it may make sense to others). Overall I think this is an incredibly simple lexer so to speak and I don't see having regex as being a good thing moving forward. We may end up with rules as we evolve identifier that may be difficult to represent in a regex and that is why I don't see value in regex, in other words, both because our solution is super simple now and because long run it may only add limitations.
I also think the non regex solution is much easier to read for me (again it's me).
Anyway as I said I am ok moving forward this way so let me know if you want a stamp here and we can move forward that way.

Reply:
I can relate to @sunshowers's rationale though. In general if you want parsing to be done in linear time (and better, being a single pass and constant, close to minimum time on each character input) you naturally think of DFAs, which can be built from regular expressions.

This particular case however, is simple enough to come up with a hand written solution that also has all the advantages above. (Or in some sense, easy enough to build the DFA manually, skipping the regex to NFA to DFA conversion.)

Reply:
OK, switched to this.

Note that I needed the regex for proptests anyway, so it isn't going away :)